Add set-region command

stephanLandwehr opened the pull request:
In case a user uses the `current-shift` command, it might be possible that
the tool has no reference what the current shift is based on the region. To
allow the user to explicitly set a region, a convenience command should be
added to the tool.

Add function to update config file with new data.

Add detection to guess the current shift based on timezone.

Update `current-shift` command to include helper message on how to
configure the current user region.

stephanLandwehr committed:
Add set-region command

In case a user uses the `current-shift` commmand, it might be possible that
the tool has no reference what the current shift is based on the region. To
allow the user to explicitly set a region, a convenience command should be
added to the tool.

Add function to update config file with new data.

Add detection to guess the current shift based on timezone.

Update `current-shift` command to include helper message on how to
configure the current user region.

stephanLandwehr committed:
Merge branch 'main' into add/set-region-command

stephanLandwehr committed:
Refactor set-region command code

Introduce constant for command name.

Introduce ShiftTime type to simplify code.

HeavyWombat commented on internal/pd/models.go:
```diff
@@ -20,15 +20,54 @@
 }
+
+//ChangeYAMLFile changes a specific value in the .pd.yml file
+func ChangeYAMLFile(name string, newValue string) error {
```
The variable `name` seems to be unused if I am not mistaken?

HeavyWombat commented on internal/pd/models.go:
```diff
@@ -20,15 +20,54 @@
 	} `yaml:"shift-times"`
 }
+
+//ChangeYAMLFile changes a specific value in the .pd.yml file
```
```suggestion
// ChangeYAMLFile changes a specific value in the .pd.yml file
```

HeavyWombat commented on internal/pd/models.go:
```diff
@@ -20,15 +20,54 @@
+//ChangeYAMLFile changes a specific value in the .pd.yml file
+func ChangeYAMLFile(name string, newValue string) error {
+
+	home, err := homedir.Dir()
+	if err != nil {
+		return err
+	}
+
+	data, err := ioutil.ReadFile(filepath.Join(home, ".pd.yml"))
+	if err != nil {
+		return err
+	}
```
I think we use this at another part of the tool as well. Can we make sure not to duplicate code and maybe introduce a common function to return the full name of the configuration file?

stephanLandwehr replied:
We do. The next commit will contain a new function for that but this part won't be changed, because the variable home is used in line 67 and therefore needed.

HeavyWombat commented on internal/cmd/set-region.go:
```diff
@@ -0,0 +1,89 @@
+
+		if len(args) == 0 {
+			ownShift, err := pd.GetProbablyOwnShift()
+			if err != nil || ownShift.Name == "" {
```
I am curious, why is `err != nil` not enough?

stephanLandwehr replied:
Not 100% sure, but if I remember correctly, if the .pd.yml file contains "own-shift:" but own-shift has not been set yet, no error occurs, but the returned name is "".

HeavyWombat commented on internal/cmd/current-shift.go:
```diff
@@ -34,17 +34,35 @@ var currentShiftCmd = &cobra.Command{
+		bunt.Printf("The next shift will be SkyBlue{%s} in %d:%02d hours\n", nextShift.Name, timeUntilNextShift/60, timeUntilNextShift%60)
+
+		if ownShiftPos == -1 {
+			bunt.Printf("Your region has not been set yet. Please run the 'LightSlateGray{pd %s [region-name]}' command\n\n", cmdName)
```
```suggestion
			bunt.Printf("\nYour region has not been set yet. In case you want to set it in the configuration, please run the 'LightSlateGray{pd %s [region-name]}'\n\n", cmdName)
```

HeavyWombat commented on internal/pd/get-shifts.go:
```diff
@@ -22,96 +22,164 @@ package pd
-// GetNextShift returns information about the next shift and time until it starts
-func GetNextShift(shifts []Shift, shiftPos int) (Shift, int, error) {
+// ShiftTime stores the amount of minutes that passed since midnight or the length of a time period in minutes
+//  3:30 am would be stored as 210 (3 * 60 + 30) and a time period of 6:15 h would be stored as 375 (6 * 60 + 15)
```
```suggestion
// 3:30 am would be stored as 210 (3 * 60 + 30) and a time period of 6:15 h would be stored as 375 (6 * 60 + 15)
```

HeavyWombat commented on internal/pd/get-shifts.go:
```diff
@@ -22,96 +22,164 @@ package pd
 
-// GetCurrentShift returns all shifts in a slice and the position of the current shift
-func GetCurrentShift() ([]Shift, int, error) {
+// GetCurrentAndOwnShift returns all shifts in a slice and the position of the current shift
+func GetCurrentAndOwnShift() ([]Shift, int, int, error) { // first int is the position of current-shift in []Shift and second int is the position of own-shift
```
```suggestion
// GetCurrentAndOwnShift returns all shifts in a slice, the position of the current shift, and
// the position of your own-shift, or an error otherwise
func GetCurrentAndOwnShift() ([]Shift, int, int, error) {
```

HeavyWombat reviewed (changes requested):
A couple of questions and some minor suggestions to change comments.

stephanLandwehr committed:
Refactor list-on-calls code

Add new func getDataFromYAMLFile.

Change func ChangeYAMLFile to be able to modify every entry in the .pd.yml file.

HeavyWombat reviewed (approved):
all good
